File: api/models/cycles.go
package models

import "time"

// CycleWorkout is one workout in a cycle's ordered sequence, carrying the
// periodization prescription: which week/phase it belongs to, how often, at what
// effort, and the readiness condition that gates advancing to the next block. It
// embeds the workout's name and theme (denormalized on read) so a cycle renders in
// a single GET, mirroring how a WorkoutMovement embeds its movement. Every
// prescription field is nullable — a mobility block may set none of them.
type CycleWorkout struct {
	Week         *int    `json:"week"`
	Phase        *string `json:"phase"`
	Frequency    *string `json:"frequency"`
	Intensity    *string `json:"intensity"`
	Conditions   *string `json:"conditions"`
	WorkoutTheme *string `json:"workout_theme"`
	WorkoutName  string  `json:"workout_name"`
	ID           int64   `json:"id"`
	WorkoutID    int64   `json:"workout_id"`
	Position     int     `json:"position"`
}

// CycleWorkoutInput is the write shape for one entry — used both when creating a
// cycle with its full sequence in one call (seed/import) and when adding a single
// workout to an existing cycle. Position is implied by array order on create and
// assigned server-side (append) on a single add.
type CycleWorkoutInput struct {
	Week       *int    `json:"week"`
	Phase      *string `json:"phase"`
	Frequency  *string `json:"frequency"`
	Intensity  *string `json:"intensity"`
	Conditions *string `json:"conditions"`
	WorkoutID  int64   `json:"workout_id"`
}

// CycleWorkoutUpdate is a partial update of one entry: a nil field is left
// unchanged. WorkoutID re-points the entry at a different workout — the swap —
// carrying the existing prescription (week/phase/frequency/...) to the substitute.
type CycleWorkoutUpdate struct {
	WorkoutID  *int64  `json:"workout_id"`
	Week       *int    `json:"week"`
	Phase      *string `json:"phase"`
	Frequency  *string `json:"frequency"`
	Intensity  *string `json:"intensity"`
	Conditions *string `json:"conditions"`
}

// CycleWorkoutOrder is the reorder body: the entry ids in their desired order. The
// repository rewrites positions to 1..n from this list.
type CycleWorkoutOrder struct {
	EntryIDs []int64 `json:"entry_ids"`
}

// Cycle is the full DB-owned record: a mesocycle — an ordered sequence of workouts
// toward a goal. TargetMetric optionally names a metric_definition so the goal is a
// real tracked number; TargetValue is the number to reach. Both dates are nullable
// so a planned cycle can be drafted before it is scheduled. Workouts is embedded on
// read (empty in list responses, populated on detail), mirroring a workout's
// movements. Dates are calendar dates on the wire ("2006-01-02").
type Cycle struct {
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	TargetMetric *string        `json:"target_metric"`
	TargetValue  *float64       `json:"target_value"`
	TargetDate   *string        `json:"target_date"`
	StartDate    *string        `json:"start_date"`
	Name         string         `json:"name"`
	GoalSummary  string         `json:"goal_summary"`
	Status       string         `json:"status"`
	Notes        string         `json:"notes"`
	Workouts     []CycleWorkout `json:"workouts"`
	ID           int64          `json:"id"`
}

// CycleCreate is the create body. Name is required (the handler validates); Status
// defaults to "planned" server-side when empty. Workouts may be supplied to build a
// complete cycle in one call (the seed/import path) or omitted and added later via
// the sub-resource.
type CycleCreate struct {
	TargetMetric *string             `json:"target_metric"`
	TargetValue  *float64            `json:"target_value"`
	TargetDate   *string             `json:"target_date"`
	StartDate    *string             `json:"start_date"`
	Name         string              `json:"name"`
	GoalSummary  string              `json:"goal_summary"`
	Status       string              `json:"status"`
	Notes        string              `json:"notes"`
	Workouts     []CycleWorkoutInput `json:"workouts"`
}

// CycleUpdate is a partial update of cycle-level fields; the workout sequence is
// managed through the sub-resource endpoints, not here. A nil pointer means "leave
// unchanged". For a nullable field (dates, target metric/value), an explicit empty
// value clears it while an absent field leaves it untouched.
type CycleUpdate struct {
	Name         *string  `json:"name"`
	GoalSummary  *string  `json:"goal_summary"`
	TargetMetric *string  `json:"target_metric"`
	TargetValue  *float64 `json:"target_value"`
	TargetDate   *string  `json:"target_date"`
	StartDate    *string  `json:"start_date"`
	Status       *string  `json:"status"`
	Notes        *string  `json:"notes"`
}

// CycleFilter carries the optional list-endpoint query params. Status scopes to one
// lifecycle state (planned|active|paused|completed); Search matches name/goal.
type CycleFilter struct {
	Status string
	Search string
}

File: web/src/api/cycles.ts
// Typed access to the cycles endpoints and their sequence sub-resource, plus the
// review capstone read. Types mirror the Go API's JSON wire contract; the DB owns id
// and timestamps.
import { http } from './client'
import type { Measurement } from './measurements'
import type { LogEntry } from './log'
import type { Session } from './sessions'

export type CycleStatus = 'planned' | 'active' | 'paused' | 'completed'

export const CYCLE_STATUSES: CycleStatus[] = ['planned', 'active', 'paused', 'completed']

// CycleWorkout is one entry in a cycle's ordered sequence, with the workout's
// name/theme embedded for render. The periodization fields are nullable.
export interface CycleWorkout {
  id: number
  workout_id: number
  workout_name: string
  workout_theme: string | null
  position: number
  week: number | null
  phase: string | null
  frequency: string | null
  intensity: string | null
  conditions: string | null
}

export interface Cycle {
  id: number
  name: string
  goal_summary: string
  target_metric: string | null
  target_value: number | null
  target_date: string | null
  start_date: string | null
  status: CycleStatus
  notes: string
  workouts: CycleWorkout[]
  created_at: string
  updated_at: string
}

// CycleWrite is the create/update payload for cycle-level fields. Fields omitted from
// an update are left unchanged; an explicit empty date/metric clears it.
export interface CycleWrite {
  name?: string
  goal_summary?: string
  target_metric?: string | null
  target_value?: number | null
  target_date?: string | null
  start_date?: string | null
  status?: CycleStatus
  notes?: string
}

// CycleWorkoutInput is the write shape for adding an entry to the sequence.
export interface CycleWorkoutInput {
  workout_id: number
  week?: number | null
  phase?: string | null
  frequency?: string | null
  intensity?: string | null
  conditions?: string | null
}

// CycleWorkoutPatch edits or swaps one entry. workout_id re-points the entry (the
// swap), carrying the untouched prescription to the substitute.
export interface CycleWorkoutPatch {
  workout_id?: number
  week?: number | null
  phase?: string | null
  frequency?: string | null
  intensity?: string | null
  conditions?: string | null
}

export interface CycleFilter {
  status?: CycleStatus
  search?: string
}

function queryString(filter: CycleFilter): string {
  const params = new URLSearchParams()
  if (filter.status) params.set('status', filter.status)
  if (filter.search) params.set('search', filter.search)
  const q = params.toString()
  return q ? `?${q}` : ''
}

export const cyclesApi = {
  list: (filter: CycleFilter = {}) => http.get<Cycle[]>(`/cycles${queryString(filter)}`),
  get: (id: number) => http.get<Cycle>(`/cycles/${id}`),
  create: (body: CycleWrite & { workouts?: CycleWorkoutInput[] }) => http.post<Cycle>('/cycles', body),
  update: (id: number, body: CycleWrite) => http.put<Cycle>(`/cycles/${id}`, body),
  remove: (id: number) => http.del(`/cycles/${id}`),
  // Sequence sub-resource — each write returns the refreshed cycle.
  addWorkout: (id: number, body: CycleWorkoutInput) => http.post<Cycle>(`/cycles/${id}/workouts`, body),
  updateWorkout: (id: number, entryId: number, body: CycleWorkoutPatch) =>
    http.patch<Cycle>(`/cycles/${id}/workouts/${entryId}`, body),
  reorderWorkouts: (id: number, entryIds: number[]) =>
    http.patch<Cycle>(`/cycles/${id}/workouts`, { entry_ids: entryIds }),
  removeWorkout: (id: number, entryId: number) => http.delData<Cycle>(`/cycles/${id}/workouts/${entryId}`),
}

// Review is the capstone read: active cycles plus recent sessions, measurements, and
// log entries in one window, the substrate for drafting the next cycle with Claude.
export interface Review {
  since: string
  active_cycles: Cycle[]
  sessions: Session[]
  measurements: Measurement[]
  log_entries: LogEntry[]
}

export const reviewApi = {
  get: (since?: string) => {
    const q = since ? `?${new URLSearchParams({ since }).toString()}` : ''
    return http.get<Review>(`/review${q}`)
  },
}

// cycleTargetSummary renders a cycle's metric target compactly (e.g.
// "deadlift-working-weight → 315"), or an empty string when there is no numeric target.
export function cycleTargetSummary(c: Cycle): string {
  if (!c.target_metric) return ''
  if (c.target_value == null) return c.target_metric
  return `${c.target_metric} → ${c.target_value}`
}

// cyclePrescriptionSummary renders an entry's week/phase/frequency/intensity compactly.
export function cyclePrescriptionSummary(cw: CycleWorkout): string {
  const parts: string[] = []
  if (cw.week != null) parts.push(`wk ${cw.week}`)
  if (cw.phase) parts.push(cw.phase)
  if (cw.frequency) parts.push(cw.frequency)
  if (cw.intensity) parts.push(cw.intensity)
  return parts.join(' · ')
}
